pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -33,36 +35,128 @@ import (
 	"k8s.io/apimachinery/pkg/watch"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
 )
 
 const resyncPeriod = 60 * time.Minute
 
+func cleanupNPLAnnotationForPod(kubeClient clientset.Interface, pod *corev1.Pod) error {
+	_, ok := pod.Annotations[k8s.NPLAnnotationStr]
+	if !ok {
+		return nil
+	}
+	delete(pod.Annotations, k8s.NPLAnnotationStr)
+	if _, err := kubeClient.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("unable to update Annotation for Pod %s/%s, %s", pod.Namespace,
+			pod.Name, err.Error())
+	}
+	return nil
+}
+
+func addRulesForNPLPorts(portTable *portcache.PortTable, allNPLPorts []rules.PodNodePort) error {
+	for _, nplPort := range allNPLPorts {
+		portTable.AddUpdateEntry(nplPort.NodePort, nplPort.PodPort, nplPort.PodIP)
+	}
+
+	if err := portTable.PodPortRules.AddAllRules(allNPLPorts); err != nil {
+		return nil
+	}
+	return nil
+}
+
+// getPodsAndGenRules fetches all the Pods on this Node and looks for valid NPL Annotations, if they
+// exist with a valid Node Port, it adds the Node port to the port table and rules. If the Node port
+// is invalid or the NPL Annotation is invalid, the NPL Annotation is removed. The Pod event handlers
+// take care of allocating a new Node port if required.
+func getPodsAndGenRules(kubeClient clientset.Interface, portTable *portcache.PortTable,
+	nodeName string) error {
+	podList, err := kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
+	if err != nil {
+		return fmt.Errorf("error in fetching the Pods for Node %s: %s", nodeName, err.Error())
+	}
+
+	allNPLPorts := []rules.PodNodePort{}
+	for _, pod := range podList.Items {
+		// For each Pod:
+		// check if a valid NPL Annotation exists for this Pod:
+		//   if yes, verifiy validity of the Node port, update the port table and add a rule to the
+		//   rules buffer.
+		podCopy := pod.DeepCopy()
```

Comment:
Move it to line 95? And if we update only annotations, probably do not copy the whole Pod.

Reply:
Moved the copying line. Thanks!
Regarding the annotations, I have commented below.